Fetch topics on demand to prevent heavy load on /topics

preetmishra opened the pull request:
This is the v1 that we discussed in [zulip-terminal > Topics on demand](https://chat.zulip.org/#narrow/stream/206-zulip-terminal/topic/Topics.20on.20demand/near/967222).

Now, topics are only fetched when they required for the `topics_view()` (for the topic toggle) or for the `_update_topic_index()`.

Tests amended with a few spelling corrections.

Fixes #493.

neiljp commented on zulipterminal/ui_tools/views.py:
```diff
@@ -767,6 +767,11 @@ def streams_view(self) -> Any:
     def topics_view(self, stream_button: Any) -> Any:
         stream_id = stream_button.stream_id
+
+        # Fetch topics and index them if they have not been fetched yet.
+        if not self.model.index['topics'][stream_id]:
+            self.model.get_topics_in_stream([stream_id])
```
This looks very similar to the code added in the model, and this seems like a good opportunity to limit direct access to the index via a model method. Here we access the updated data below in any case.

neiljp commented on zulipterminal/model.py:
```diff
@@ -853,6 +827,11 @@ def _update_topic_index(self, stream_id: int, topic_name: str) -> None:
+        # Fetch topics before updating the index if they have not been fetched
+        # yet via the topic toggle.
+        if not self.index['topics'][stream_id]:
+            self.get_topics_in_stream([stream_id])
```
This is not specific to your change here, but reading this code reminds me that `get_` is unclear just from the name - this is actually an API call, whereas `get_` can express any case where you're gathering data through an object, from a resource or otherwise (particularly if you come from languages where "getters" and "setters" are standard). In this context I might expect any external access of the topic list for a stream to be hidden behind a `get_*` (or no prefix), with any potential API call (`fetch_*`) generally happening behind the scenes.

To clarify, this is really another issue, since we do this in a few cases.

preetmishra replied:
Agreed. It would make it easier to depict what the method actually does.

preetmishra commented:
@neiljp Thanks for the review. I have updated the PR to address your suggestions.

neiljp commented on zulipterminal/model.py:
```diff
@@ -853,6 +835,10 @@ def _update_topic_index(self, stream_id: int, topic_name: str) -> None:
         """
+        # Fetch and index topics before updating the index if they have not
+        # been fetched yet via the topic toggle.
+        self.set_initial_topics_index(stream_id)
+
         topic_index = self.index['topics'][stream_id]
```
Using a method in this iteration is a good first step, but note that in both cases the code is setting the initial index - ensuring it exists or is fetched - then looking it up for a particular stream. This is a standard pattern where one object handles caching/synchronizing internally, but the external users of that code don't need to consider how that happens, or even be aware of it - they just want the list of topics! Of course, there can be cases where the external code might want this (eg. observer pattern) but we don't need it here, at least yet.

preetmishra commented:
@neiljp Thanks for the exposition. :+1: I have updated the commit to include the lookup and limit access to `index`. I have also added another commit to encapsulate `get_topics_in_stream()`.

preetmishra commented:
@neiljp Thanks for further comments. I have updated the PR based on your suggestion and our discussion in PM.

neiljp commented on zulipterminal/model.py:
```diff
@@ -841,14 +841,17 @@ def _update_topic_index(self, stream_id: int, topic_name: str) -> None:
+        for topic_iterator, topic in enumerate(topic_list):
             if topic == topic_name:
-                topic_index.insert(0, topic_index.pop(topic_iterator))
+                topic_list.insert(0, topic_list.pop(topic_iterator))
                 return
```
This doesn't update the data now - that's what the code says, but it took a while to figure out why. The UI still works since we update them independently. We should probably comment in the caller that it works this way, and whether we want to determine if if it's important to keep this synchronized, or how.

preetmishra commented:
@neiljp Thanks for catching the synchronization bit. I have updated the PR to address that and other in-line suggestions.

preetmishra commented:
Updated with an identity assert for `topics_in_stream()`.

neiljp commented:
@preetmishra Thanks for this contribution - this closes an important bug from the perspective of the server :+1:

@timabbott I mentioned this briefly on czo (#z-t>topics on demand), but now that this approach is complete, what do you think about a slower hybrid approach where topics are loaded slowly in the background?

timabbott commented:
I think there's probably a broader API discussion we should have about whether the server should offer a "Recent topics for all streams" API feature in /feature; I think we're going to end up needing something like that for the "Recent topics" feature in the webapp.  @amanagr FYI.
